### ISSUE TITLE
[Pokemon Emerald] issue and fix and other possible fixes

### ISSUE THREAD

**Issue by cotodevel** [Gericom/GBARunner2#52]
Since I am no longer coding/maintaining gbaemu4ds I wanted to share some tips I have found while at it and I am sure those will be of help to GBARunner2 project. 

1#)
GBARunner2: Since 2017 or so I was never able to run Pokémon Emerald (U) again. So I decided to build a "custom" GBARunner2 version having all the latest improvements + my own. But I found the game broke between 2017 - 2018 causing bios intro loops or whitescreens.

Turns out a certain commit broke it (and possibly other games as well)

https://github.com/Gericom/GBARunner2/commit/216734de3e1f09e476657568589d1a5dd01c38d1 -> exact commit which breaks pokemon emerald (U)

https://github.com/Gericom/GBARunner2/commit/a61e70e66c517c49d229e31e7cd81a93418b60c3
-> pokemon emerald (U) still works (one commit right before the above one)

Current Environment I have: devkitARM r45 and the same libnds and libfat GBATemp user mentioned here: https://gbatemp.net/threads/gbarunner2.451970/page-75#post-8672889

Also when I added stuff to gbaemu4ds, I found pokemon games to break often when VRAM or IO read / writes were unhandled / omitted or deadlocks when certain load/store memory operations would leave bus conflicts unresolved between different memory such as the VRAM and the EWRAM (at ARM9), causing hangs. That takes me to point 2).

2#
Wario Land 4 freezes:

I enabled PU interrupts for any kind of read / writes at 0x06000000 so I could handle 8 bit writes on ARM9-mode WRAM (I know you can patch SDK memcpy but I still wanted to give it a try), and I was able to reproduce the Wario Land 4 issue where the character gets stuck right before jumping into a portal.

Basically there´s a bus conflict between EWRAM and VRAM when moving data and then a MPU interrupt happens (that is if the MPU steps over the latch signal the VRAM hooks on). If I omit entirely the MPU to watch VRAM writes (of which GBA hypervisor does natively) then the timing deadlock goes away and the game proceeds as normal. I can provide the sources and a commit so you can replicate that by yourself. 

edit: Also I think running code from VRAM at ARM9 is likely to cause similar issues.

**Comment by Gericom** [comment]
Problem number 1 is most likely related to code above the first 2 MB which is currently loaded. Before that one commit, more data was loaded. I might look into adjusting that at some point.

As for problem 2, are you saying that basically the correct emulation of 8 bit writes to vram and thus the requirement to disallow direct writing causes the timing problem? Running code from vram is supported btw.

**Comment by cotodevel** [comment]
"What kind of bus conflicts do you mean btw? A timing thing?"
A race condition unresolved by the ARM AHB bus manager. Let´s say the NDS hardware design is by no means lockstep-safe. I can enumerate the issues I had by trying to relocate code to Shared WRAM and then make executable code from there to try to access ITCM/DTCM... these are simply ignored. That could be because the TCM, the caches and the MPU is hooked directly into the ARM processor and the ARM processor must assert read/writes from these or from the AHB slave bus.

During the years I have found that any ITCM/DTCM / MPU -> ARM -> EWRAM (or other slave AHB device such as Shared WRAM, VRAM, ITCM, MPU, caches) allows to lock-unlock "mutexes" accesses if race conditions ever happen between slave AHB devices. But other devices might not be able to do that by themselves just by being wired onto the AHB bus. 

And yes, using the MPU to take over any AHB slave device other than the EWRAM is likely to cause bus conflicts. (which excludes the TCM, caches)

**Comment by Gericom** [comment]
Hmm, I guess it's basically some kind of priority thing. Note that itcm/dtcm and cache is not hooked to the bus.

**Comment by cotodevel** [comment]
Sent you a PM in GBATemp. Closing issue... ; - )

**Comment by qamichele59** [comment]
#52
